Make source-based code coverage compatible with MIR inlining

tmiasko opened the pull request:
When codegenning code coverage use the instance that coverage data was
originally generated for, to ensure basic level of compatibility with
MIR inlining.

Fixes #83061

richkadel commented:
Nice! I didn't realize you could still get the instance from the `source_scope`, but that definitely simplified things over what I thought was needed per #83061 .

Thank you @tmiasko !

Are you willing to add some tests in `src/test/run-make-fulldeps/coverage/assert.rs` or elsewhere?

richkadel commented:
Oh, this test also should be removed (now that the warning is deleted):

src/test/ui/mir/mir-inlining/inline-instrument-coverage-fail.rs

tmiasko commented:
Thanks. I removed now obsolete test for incompatibility, I also adjusted the computation of reachable functions to account for situation where function does not have its code generated directly, but is still reachable after being inlined into another. I will look into adding some test cases that combine code coverage and inlining.

richkadel commented:
BTW I can do another review when done, but I don't have Compiler Team approval privileges, so you'll need someone to do that (likely @wesleywiser or @tmandry) to push to bors, when ready.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2117889586) [(plain)](https://github.com/rust-lang/rust/commit/7d005914b2b08ec8efb88e243ad7b782e6e3c567/checks/2117889586/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Checking which error codes lack tests...
Found 435 error codes
Found 0 error codes with no tests
Done!
tidy error: /checkout/src/test/run-make-fulldeps/coverage/inline.rs: too many trailing newlines (2)
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/tidy
Build completed unsuccessfully in 0:00:14
```

</details>

tmiasko commented:
Added support for compile-flags to coverage tests and used it to create a MIR inlining test case.

wesleywiser commented on compiler/rustc_codegen_ssa/src/coverageinfo/map.rs:
```diff
@@ -94,9 +96,18 @@ impl<'tcx> FunctionCoverage<'tcx> {
-        self.expressions[expression_index]
-            .replace(Expression { lhs, op, rhs, region })
-            .expect_none("add_counter_expression called with duplicate `id_descending_from_max`");
+        if let Some(previous_expression) = self.expressions[expression_index].replace(Expression {
+            lhs,
+            op,
+            rhs,
+            region: region.clone(),
```
Nit: Might be slightly nicer to extract a local for `Expression { lhs, op, rhs ... }` and then use it in both places instead of constructing twice.

wesleywiser reviewed (approved):
This is great!

wesleywiser commented:
@bors r+

bors commented:
:pushpin: Commit 0d84e0b68c78006c4aa6f5ec021d92277be7d3cc has been approved by `wesleywiser`

<!-- @bors r=wesleywiser 0d84e0b68c78006c4aa6f5ec021d92277be7d3cc -->
<!-- homu: {"type":"Approved","sha":"0d84e0b68c78006c4aa6f5ec021d92277be7d3cc","approver":"wesleywiser"} -->
